community/kernel/src/test/java/org/neo4j/kernel/impl/locking/RelationshipCreateDeleteLockOrderingIT.java

```diff
@@ -0,0 +1,112 @@
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.kernel.impl.MyRelTypes;
+import org.neo4j.test.DatabaseRule;
+import org.neo4j.test.ImpermanentDatabaseRule;
+import org.neo4j.test.Race;
+import org.neo4j.test.RandomRule;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Testing on database level that creating and deleting relationships over the the same two nodes
+ * doesn't create unnecessary deadlock scenarios, i.e. that locking order is predictable and symmetrical
+ */
+public class RelationshipCreateDeleteLockOrderingIT
+{
+    @Rule
+    public final DatabaseRule db = new ImpermanentDatabaseRule();
+    @Rule
+    public final RandomRule random = new RandomRule();
+
+    @Test
+    public void shouldNotDeadlockWhenConcurrentCreateAndDeleteRelationships() throws Throwable
+    {
+        // GIVEN (A) -[R]-> (B)
+        final Node a;
+        final Node b;
+        try ( Transaction tx = db.beginTx() )
+        {
+            (a = db.createNode()).createRelationshipTo( b = db.createNode(), MyRelTypes.TEST );
+            tx.success();
+        }
+
+        // WHEN
+        Race race = new Race();
+        // a bunch of deleters
+        for ( int i = 0; i < 30; i++ )
+        {
+            race.addContestant( new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try ( Transaction tx = db.beginTx() )
+                    {
+                        Node node = random.nextBoolean() ? a : b;
+                        for ( Relationship relationship : node.getRelationships() )
+                        {
+                            try
+                            {
+                                relationship.delete();
+                            }
+                            catch ( IllegalStateException e )
+                            {
+                                // This is OK and expected since there are multiple threads deleting
+                                assertTrue( e.getMessage().contains( "already deleted" ) );
+                            }
+                        }
+                        tx.success();
+                    }
+                }
+            } );
+        }
+
+        // a bunch of creators
+        for ( int i = 0; i < 30; i++ )
+        {
+            race.addContestant( new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try ( Transaction tx = db.beginTx() )
+                    {
+                        boolean order = random.nextBoolean();
+                        Node start = order ? a : b;
+                        Node end = order ? b : a;
+                        start.createRelationshipTo( end, MyRelTypes.TEST );
+                        tx.success();
+                    }
+                }
+            } );
+        }
+
+        // THEN there should be no thread throwing exception, especially DeadlockDetectedException
```

Comment:
Actually, that's another thing entirely and this test will not run with forseti anyway